src/NzbDrone.Core/MediaFiles/MediaInfo/MediaInfoUpdater.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.IO;
+using NLog;
+using NzbDrone.Common.Disk;
+using NzbDrone.Core.Tv;
+
+namespace NzbDrone.Core.MediaFiles.MediaInfo
+{
+    public interface IUpdateMediaInfo
+    {
+        void Update(EpisodeFile mediaFile, Series series);
+        EpisodeFile GetOrUpdate(EpisodeFile mediaFile, Series series);
+    }
+
+    public class MediaInfoUpdater : IUpdateMediaInfo
+    {
+        private readonly IVideoFileInfoReader _videoFileInfoReader;
+        private readonly IDiskProvider _diskProvider;
+        private readonly IMediaFileService _mediaFileService;
+        private readonly Logger _logger;
+
+        public MediaInfoUpdater(Logger logger, IMediaFileService mediaFileService, IDiskProvider diskProvider, IVideoFileInfoReader videoFileInfoReader)
+        {
+            _logger = logger;
+            _mediaFileService = mediaFileService;
+            _diskProvider = diskProvider;
+            _videoFileInfoReader = videoFileInfoReader;
+        }
+
+        public void Update(EpisodeFile mediaFile, Series series)
+        {
+            if (string.IsNullOrEmpty(series.Path) || string.IsNullOrEmpty(mediaFile.RelativePath)) return;
+
+            var path = Path.Combine(series.Path, mediaFile.RelativePath);
+            Update(mediaFile, path);
+        }
+
+        public EpisodeFile GetOrUpdate(EpisodeFile mediaFile, Series series)
+        {
+            if(mediaFile.MediaInfo.SchemaRevision < VideoFileInfoReader.CURRENT_MEDIA_INFO_SCHEMA_REVISION)
```

Comment:
I think you misunderstood this part of the exercise. We already do this based on MINIMUM_MEDIA_INFO_SCHEMA_REVISION in UpdateMediaInfoService. This is because we don't want to update mediainfo for the entire library unless it's absolutely necessary.
The idea is that for HDR (and _only_ HDR) we need at least 5.

So basically, if the user has {MediaInfo VideoDynamicRange} in their format, we need to ensure the file is v5.
We can do that two ways, either make the  lazy loaded and when called ensure that we have the right schema version. Or, do a second (initial) pass using the regex to determine the minimum schema version required.

Reply:
Yes. I extracted the update part of the code from `UpdateMediaInfoService` into that class, and it's used in the `{MediaInfo VideoDynamicRange}` token handler, as well as the `UpdateMediaInfoService`. As I understood it, the update service gets called for a full show scan and updates all episodes that don't have the current revision. The token handler will call GetOrUpdate and update the mediainfo if it's not the current revision also. I thought that the token handler will only run if the token is in the naming config.

Reply:
> I thought that the token handler will only run if the token is in the naming config.

It will, but then you should provide the expected version number in GetOrUpdate, not via CURRENT_*. If another unrelated change is made in the future as v6, then you don't want the entire library to be rescanned just because ppl have the HDR token.
It also doesn't need to run on AudioCodec... yes there's a difference due to the new audio codec support added, but forcing a refresh there would cause a full library scan for most users. Whereas using the HDR token is a new one that the user would have to add explicitly.

_PS: To explain why it's so important to avoid rescanning: There are quite a bit of users with cloud storage and several cloud providers do not support partial file reads. Rescanning all mediainfo would cause a lot of traffic and therefore be rather undesirable. And the acceptable trade-off is to only scan mediainfo for new imports._

Reply:
ahh ok, yes I was wondering if each individual media info formatter should know about its own revision. so yes, HDR is >=v5 and some other property is >=v6

Yep I understand now, thanks for explaining the issues around rescanning, I'll bear that in mind. I didn't consider cloud storage. Makes sense to remove the audio codec ones too considering that. It'll use the new audio stuff if its there.

Reply:
I put audio codec back to how it was and changed the video dynamic range to specify its minimum required revision. An alternative here would be to convert all the formatters to specify their minimum required revision.

Also it might be better to change the formatter from a static to an injectable class that can depend on the media info updater and manage updating itself, rather than passing in the func to each format method?